File: tests/termviz/test_plan_workspace.py
"""Real compiler and native PTY: revisions, diagnostics and reversible layouts."""
import json
import os
from pathlib import Path
import shutil
import subprocess
import tempfile
import time

from pty_support import Session

ROOT = Path(__file__).resolve().parents[2]
BUILD = Path(os.environ.get("BUILD_DIR", ROOT / "build")).resolve()

with tempfile.TemporaryDirectory(prefix="hydra-plan-workspace-") as folder:
    base = Path(folder)
    repo = base / "repo"
    shutil.copytree(ROOT / "tests/fixtures/plan/repo", repo)
    draft, policy = base / "draft.json", base / "policy.json"
    shutil.copy(ROOT / "tests/fixtures/plan/plan.json", draft)
    shutil.copy(ROOT / "tests/fixtures/plan/policy.json", policy)
    env = {**os.environ, "TERM": "xterm-256color", "HYDRA_FLEET_BIN": str(BUILD / "hydra-fleet"), "HYDRA_HOME": str(base / "home"), "HYDRA_NONINTERACTIVE": "1", "HYDRA_SKIP_AI": "1"}
    for args in (["git", "init", "-q"], ["git", "add", "."],
                 ["git", "-c", "user.name=Test", "-c", "user.email=test@example.com", "commit", "-qm", "fixture"],
                 [str(ROOT / "bin/hydra"), "init", "--no-agent", "--trust"]):
        subprocess.run(args, cwd=repo, env=env, check=True, stdout=subprocess.DEVNULL)
    s = Session([str(BUILD / "hydra-tui"), "--hydra", str(ROOT / "bin/hydra")], 140, 40, env=env, cwd=repo)
    try:
        s.until("A CONVERSATION")
        s.send("P")
        s.until("Draft JSON path:")
        s.send(str(draft) + "\r")
        s.until("Policy JSON path:")
        s.send(str(policy) + "\r")
        s.until("Revision 1 / DRAFT")
        s.send("VB\t\t")
        s.until("READY / awaiting approval", timeout=30)
        s.send("z")
        s.until("READY / awaiting approval")
        first = s.screen.text()
        assert "Revision 1" in first
        s.send("jjjjB")
        s.until("A CONVERSATION")
        s.send("B")
        s.until("B PLAN OVERVIEW")
        assert "Revision 1" in s.screen.text()
        changed = json.loads(draft.read_text())
        changed["objective"] = "Revised workspace acceptance objective"
        draft.write_text(json.dumps(changed))
        s.until("Revision 2 / DRAFT", timeout=8)
        assert "READY / awaiting approval" not in s.screen.text()
        s.send("V")
        s.until("READY / awaiting approval", timeout=30)
        s.send("kkkk")
        s.until("Revised workspace acceptance objective")
        for width, height in [(40, 10), (80, 24), (140, 40)]:
            if width == 40:
                s.send("j")
                time.sleep(.1)  # Leave a pre-resize frame queued in the PTY.
            before_resize = len(s.raw)
            s.resize(width, height)
            s.pump(.3)
            if s.screen.overflow:
                (ROOT / "build/plan-resize-failure.ansi").write_bytes(s.raw[before_resize:])
            assert s.screen.overflow == 0, f"{width}x{height}, overflow={s.screen.overflow}, clears={s.screen.clears}\n{s.screen.text()}"
        draft.write_text('{"broken":')
        s.until("Revision 3 / DRAFT", timeout=8)
        s.send("V")
        s.until("INVALID", timeout=30)
        s.send("CA")
        s.until("A CONVERSATION")
        s.close(keys=b"q")
    finally:
        s.abort()
print("PASS plan workspace: real compilation, revised scope, invalid input, layouts and terminal restoration")

File: tests/termviz/test_plan_workspace.c
#define _XOPEN_SOURCE 700
#include "hydra_fixture.h"
#include <json-c/json.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
static struct hf_fixture f;
#define U(marker, timeout) tv_until(&s, (marker), (timeout))
int main(void) {
    struct tv_session s;
    char draft[4096], policy[4096], path[4096], text[8192];
    int sizes[][2] = {{40, 10}, {80, 24}, {140, 40}};
    size_t i;
    json_object *changed;
    tv_init();
    hf_init(&f, "hydra-plan-workspace", "repo", true, false);
    tv_format(draft, sizeof(draft), "%s/draft.json", f.base);
    tv_format(policy, sizeof(policy), "%s/policy.json", f.base);
    tv_format(path, sizeof(path), "%s/tests/fixtures/plan/plan.json", f.root);
    tv_copy(path, draft, false);
    tv_format(path, sizeof(path), "%s/tests/fixtures/plan/policy.json", f.root);
    tv_copy(path, policy, false);
    hf_commit_init(&f);
    hf_open(&f, &s);
    U("A CONVERSATION", 3);
    tv_send(&s, "P");
    U("Draft JSON path:", 3);
    tv_send(&s, draft);
    tv_send(&s, "\r");
    U("Policy JSON path:", 3);
    tv_send(&s, policy);
    tv_send(&s, "\r");
    U("Revision 1 / DRAFT", 3);
    tv_send(&s, "VB\t\t");
    U("READY / awaiting approval", 30);
    tv_send(&s, "z");
    U("READY / awaiting approval", 3);
    CHECK(tv_contains(&s, "Revision 1"), "first revision");
    tv_send(&s, "jjjjB");
    U("A CONVERSATION", 3);
    tv_send(&s, "B");
    U("B PLAN OVERVIEW", 3);
    CHECK(tv_contains(&s, "Revision 1"), "layout retains revision");
    changed = json_object_from_file(draft);
    CHECK(changed, "draft JSON");
    json_object_object_add(changed, "objective",
                           json_object_new_string("Revised workspace acceptance objective"));
    CHECK(!json_object_to_file(draft, changed), "write revised draft");
    json_object_put(changed);
    U("Revision 2 / DRAFT", 8);
    CHECK(!tv_contains(&s, "READY / awaiting approval"), "revision invalidates approval");
    tv_send(&s, "V");
    U("READY / awaiting approval", 30);
    tv_send(&s, "kkkk");
    U("Revised workspace acceptance objective", 3);
    for (i = 0; i < 3; i++) {
        if (sizes[i][0] == 40) {
            tv_send(&s, "j");
            tv_sleep(.1);
        }
        tv_resize(&s, sizes[i][0], sizes[i][1]);
        tv_pump(&s, .3);
        if (s.screen.overflow) {
            tv_format(path, sizeof(path), "%s/build/plan-resize-failure.ansi", f.root);
            tv_write(path, s.raw ? s.raw : "");
            fprintf(stderr, "%s\n", tv_text(&s));
        }
        CHECK(!s.screen.overflow, "plan resize overflow");
    }
    tv_write(draft, "{\"broken\":");
    U("Revision 3 / DRAFT", 8);
    tv_send(&s, "V");
    U("INVALID", 30);
    tv_send(&s, "CA");
    U("A CONVERSATION", 3);
    tv_close(&s, "q", 0, 0);
    tv_format(text, sizeof(text),
              "PASS plan workspace: compilation, revised scope, invalid JSON, reversible layouts "
              "and termios (%s)",
              f.base);
    puts(text);
    {
        const char *remove[] = {"rm", "-rf", f.base, NULL};
        tv_command_ok(NULL, remove);
    }
    return 0;
}
